File: algo/1717.cpp
#include <stdio.h>
using namespace std;
int p[1000001];
int find(int x);
void Union(int x, int y);
int main()
{
    int n, m;
    scanf("%d %d", &n, &m);
    for(int i=1; i<=n; i++)
        p[i] = i;
    int a, b, c;
    for(int i=0; i<m; i++)
    {
        scanf("%d %d %d", &a, &b, &c);
        if(a) // 같은 집합인지 확인
        {
            if(find(b) == find(c))
                printf("YES\n");
            else
                printf("NO\n");
        }
        else
            Union(b, c);
    }
    return 0;
}

int find(int x)
{
    if(x == p[x])
        return x;
    else
    {
        int y = find(p[x]);
        p[x] = y;
        return y;
    }   
}

void Union(int x, int y)
{
    int a = find(x);
    int b = find(y);
    if(a!=b)
        p[b] = a;
}

File: algo/1717.py
def solution(n, input_list):
    union_list =[i for i in range(n+1)]
    for i in input_list:
        if i[0] == 0:
            Union(i[1], i[2], union_list)
        elif i[0] == 1:
            if Find(i[1], union_list) == Find(i[2], union_list):
                print("YES")
            else:
                print("NO")

def Find(x, parent):
    if x == parent[x]:
        return x
    else:
        y = Find(parent[x], parent)
        parent[x] = y
        return y

def Union(x, y, parent):
    x = Find(x, parent)
    y = Find(y, parent)
    parent[x] = y

if __name__ == "__main__":
    nm = list(map(int, input().split()))
    input_list = []
    for m in range(nm[1]):
        input_list.append(list(map(int, input().split())))
    solution(nm[0], input_list)